lib/Dialect/FIRRTL/Transforms/LowerTypes.cpp

```diff
@@ -701,6 +702,139 @@ void TypeLoweringVisitor::visitExpr(SubindexOp op) {
+  while (auto multiDimOp = inputVector.getDefiningOp<SubaccessOp>()) {
+    // Iterate until the input is itself a SubAccess Op.
+    allDims.push_back(multiDimOp);
+    indicesRange.push_back(0);
+    inputVector = multiDimOp.input();
+  }
+  SmallVector<Value, 8> loweredVector;
+  // Flatten the vector and bundle type recursively and populate the
+  // loweredVector with individual elements after lowering. For example, a[2] is
+  // lowered to a_0,a_1 and  a: {data, valid}[2] is lowered to a_0_data,
+  // a_0_valid, a_1_data, a_1_valid, in that order and the order is important.
+  getAllBundleLowerings(inputVector, loweredVector);
+  // This records all the mux trees generated after lowering op. One mux tree
+  // for each bundle field element.
+  SmallVector<Value, 8> muxTreeVec(1, loweredVector[0]);
+  // This records a map of bundle subfield name and the corresponding subfield
+  // operation.
+  DenseMap<StringRef, SubfieldOp> subfieldUsers;
+  BundleType vectorOfBundleType = vectorElementType.dyn_cast<BundleType>();
+
+  if (vectorOfBundleType) {
+    // If the vector element is not a bundle, then only one mux tree generated,
+    // else one mux tree for each bundle element. This loop initializes the mux
+    // tree with the first element of the vector (for each element of the
+    // bundle).
+    for (size_t elem = 1; elem < vectorOfBundleType.getNumElements(); elem++)
+      muxTreeVec.push_back(loweredVector[elem]);
+    // User of a vector of bundles is a SubfieldOp.
+    for (auto u : op->getUsers())
+      if (auto subField = dyn_cast<SubfieldOp>(u))
+        subfieldUsers[subField.fieldname()] = subField;
+  }
+
+  // Create a true constant.
+  Value constTrue = builder->create<ConstantOp>(
+      op.getLoc(), UIntType::get(op.getContext(), 1),
+      getIntAttr(APInt(1, 1), op.getContext()));
+
+  size_t dimRange = 0;
+  for (size_t elem = muxTreeVec.size(); elem < loweredVector.size();
+       elem += muxTreeVec.size()) {
+    Value isIndexEq = constTrue;
+    for (size_t ind = 0; ind < allDims.size(); ind++) {
+      // Get the index variable.
+      auto index = allDims[ind].index();
+      // Get the constnat int, with which to compare the index.
+      auto indexInt = indicesRange[ind];
+      auto selectWidth =
+          index.getType().cast<FIRRTLType>().getBitWidthOrSentinel();
+      // Create " && index == indexInt?"
+      isIndexEq = builder->create<AndPrimOp>(
+          UIntType::get(op.getContext(), 1), isIndexEq,
+          builder->create<EQPrimOp>(
+              op.getLoc(), UIntType::get(op.getContext(), 1), index,
+              builder->create<ConstantOp>(
+                  op.getLoc(), UIntType::get(op.getContext(), selectWidth),
+                  getIntAttr(APInt(selectWidth, indexInt), op.getContext()))));
+      // This is the logic to generate the range of constant integers, with
+      // which to compare the index.
+      if (ind == dimRange) {
+        size_t maxElems = 1 << selectWidth;
```

Comment:
I think you need `1ULL << selectWidth` to avoid truncating at 32-bits, but it probably doesn't matter.